src/java.base/share/classes/sun/nio/ch/FileChannelImpl.java

```diff
@@ -534,109 +544,215 @@ public void force(boolean metaData) throws IOException {
+            assert Thread.currentThread().isInterrupted();
+            cbie = e;
+        } else if (!uninterruptible && Thread.currentThread().isInterrupted()) {
+            cbie = new ClosedByInterruptException();
+        }
+        if (cbie != null) {
+            try {
+                this.close();
+            } catch (IOException ioe) {
+                cbie.addSuppressed(ioe);
+            }
+            try {
+                other.close();
+            } catch (IOException ioe) {
+                cbie.addSuppressed(ioe);
+            }
+            return cbie;
+
+        }
+        // one of the channels was closed during the transfer
+        if (cce instanceof AsynchronousCloseException ace) {
+            return ace;
+        } else {
+            var ace = new AsynchronousCloseException();
+            ace.addSuppressed(cce);
+            return ace;
+        }
+    }
+
+    /**
+     * Transfers bytes from this channel's file to the resource that the given file
+     * descriptor is connected to.
+     */
+    private long transferToFileDescriptor(long position, int count, FileDescriptor targetFD) {
+        long n;
+        boolean append = fdAccess.getAppend(targetFD);
+        do {
+            long comp = Blocker.begin();
+            try {
+                n = nd.transferTo(fd, position, count, targetFD, append);
+            } finally {
+                Blocker.end(comp);
+            }
+        } while ((n == IOStatus.INTERRUPTED) && isOpen());
+        return n;
+    }
 
-    private long transferToDirectlyInternal(long position, int icount,
-                                            WritableByteChannel target,
-                                            FileDescriptor targetFD)
+    /**
+     * Transfers bytes from this channel's file to the given channel's file.
+     */
+    private long transferToFileChannel(long position, int count, FileChannelImpl target)
         throws IOException
     {
-        assert !nd.transferToDirectlyNeedsPositionLock() ||
-               Thread.holdsLock(positionLock);
+        if (transferToFileDirectNotSupported)
+            return IOStatus.UNSUPPORTED_CASE;
 
-        long n = -1;
-        int ti = -1;
+        final FileChannelImpl source = this;
+        boolean completed = false;
         try {
             beginBlocking();
-            ti = threads.add();
-            if (!isOpen())
-                return -1;
-            boolean append = fdAccess.getAppend(targetFD);
-            do {
-                long comp = Blocker.begin();
+            int sourceIndex = source.beforeTransfer();
+            try {
+                int targetIndex = target.beforeTransfer();
                 try {
-                    n = nd.transferTo(fd, position, icount, targetFD, append);
+                    long n = transferToFileDescriptor(position, count, target.fd);
+                    if (n == IOStatus.UNSUPPORTED_CASE) {
+                        transferToFileDirectNotSupported = true;
```

Comment:
On Linux, `transferToFileDescriptor` will fail with `UNSUPPORTED_CASE` if the target is open for appending. I don't think that this should mean that all future fast transfers to a file should be suppressed.

Reply:
I'll check it but all I've really done here is rename fileSupported to transferToFileDirectNotSupported and inverted the values.

Reply:
In that case the situation did not change and whether it poses a problem could be investigated under another issue.

Reply:
I think the simplest thing is to remove fileSupported for now.  Also if we limit the get/set of transferToDirectNotSupported to transferToDirect, and limit the get/set of transfeFromDirectNotSupported to transferFromDirect then it removes any confusion as to when the direct cases are disabled. I've done that now and updated the PR